### ISSUE TITLE
High Sierra breaks getpts

### ISSUE THREAD

**Issue by mennis** [9fans/plan9port#110]
In src/cmd/9term/bsdpty.c we exit with sysfatal("no ptys"). This effects win and 9term. 

```
int
getpts(int fd[], char *slave)
{
	char *a, *z;
	char pty[] = "/dev/ptyXX";

	for(a=abc; *a; a++)
	for(z=_123; *z; z++){
		pty[8] = *a;
		pty[9] = *z;
		if((fd[1] = open(pty, ORDWR)) < 0){
			if(errno == ENOENT)
				break;
		}else{
			fchmod(fd[1], 0620);
			strcpy(slave, pty);
			slave[5] = 't';
			if((fd[0] = open(slave, ORDWR)) >= 0)
				return 0;
			close(fd[1]);
		}
	}
	sysfatal("no ptys");
	return 0;
}
```

Debugging with print I see that for each case of the pty we attempt to open (not the slave) we get ENOENT back.  The man page has not been updated and ls /dev/pty* has the same filenames.

posix_openpt does exist in MacOS and appears to have been added in OSX 10.6.2 . Perhaps this is the way forward. It looks like the Eclipse Community ran into this as well:

https://bugs.eclipse.org/bugs/show_bug.cgi?id=519886

There it is indicated that the minor device number of ptmx has changed to 1 from previous versions where it is 4.  Nothing pty related observed in openradar.

**Comment by fjballest** [comment]
Hi, I found this works just fine.
HTH.
  -nemo.

int
getpts(int fd[], char *slave)
{
	char *a;

	fd[1] = posix_openpt(ORDWR);
	if (fd[1] < 0) {
		sysfatal("no ptys: open: %r");
	}
	a = ptsname(fd[1]);
	if (a == NULL) {
		sysfatal("no ptys: ptsname: %r");
	}
	if (grantpt(fd[1]) < 0) {
		sysfatal("no ptys: grantpt: %r");
	}
	if (unlockpt(fd[1]) < 0) {
		sysfatal("no ptys: unlockpt: %r");
	}
	strcpy(slave, a);
	if((fd[0] = open(a, ORDWR)) >= 0)
		return 0;
	sysfatal("no ptys: open slave: %r");
}

**Comment by gdiazlo** [comment]
The OpenBSD code seems to work for me:

```
cd $PLAN9/src/cmd/9term/
cp OpenBSD.c Darwin.c
sudo 9 mk install
```

**Comment by fjballest** [comment]
I think avoiding fchmod and using the posix interface is the more portable way these days, or so it seems.

**Comment by 0intro** [comment]
Thanks for your investigations.

It seems there are multiple solutions. Should we use `openpty` like OpenBSD or `posix_openpt` like Francisco suggests? I understand they both fix the issue on High Sierra, but do they work with older versions? We want to keep backward compatibility with the older Mac OS X versions.

In the worst case, we can always use a `#if OSX_VERSION >= 101300` condition.

**Comment by mennis** [comment]
I don't have access to many previous versions of  OS X directly but through travis I can see that the following build successfully:

xcode9		OS X 10.12 Sierra
xcode8.3	OS X 10.12 Sierra
xcode8.2	OS X 10.12 Sierra
xcode8.1	OS X 10.12 Sierra
xcode8		OS X 10.11 El Capitan
xcode7.3 	OS X 10.11 El Capitan
xcode6.4	OS X 10.10 Yosmite

using `openpty`. If it is to be believed that building means it works then this should cover many users.  A proper test would probably be to have a test executable built and then run in the build process as a [stage](https://docs.travis-ci.com/user/build-stages/define-steps/).

`posix_openpt` being around since  10.6 Snow Leopard (2011) seems to indicate to me that moving the code to use `posix_openpt` might reduce the number of implementations.

Others got `posix_openpt` around:

NetBSD 3.0 2005
OpenBSD 5.3  2013
DragonFly 2.4 2009

Glibc 2.2.1 2001

**Comment by 0intro** [comment]
I've committed the change to use openpty on macOS, using the same code as OpenBSD. It should fix the getpts issue on macOS 10.13 without breaking compatibility on older Mac OS X versions.

Maybe in the future we could use posix_openpt, like Francisco suggested.
